[1.7]Lua expressions in damage  formula

kdmcser opened the pull request:
make modder can define damage formula in gameConfig.json. 
the performance is 3~4 μs per call, so it will not make AI turns very long.

kdmcser commented:
unable to solve android dependencies

kambala-decapitator commented:
disable android temporarily on CI or set `fail-fast: false` that other platforms don't stop building, we'll merge without it and provide LuaJIT later. IIRC same issue will be with Apple and mingw.

kambala-decapitator commented:
you can also hide new feature under "feature flag" (a define basically) that other platforms don't fail

kdmcser commented:
use #ifdef in the code？

kambala-decapitator commented:
> use #ifdef in the code？

Yep

IvanSavenko commented on lib/battle/DamageCalculator.h:
```diff
@@ -25,8 +26,10 @@ class DLL_LINKAGE DamageCalculator
 {
 	const CBattleInfoCallback & callback;
 	const BattleAttackInfo & info;
+    static std::unique_ptr<LuaExpressionEvaluator> attackSkillEvaluator;
+    static std::unique_ptr<LuaExpressionEvaluator> defenseSkillEvaluator;
```
Static lua states will not work when accessed from multiple threads. And BattleAI does uses multithreading. And creating lua state every time would be too costly...

Need to investigate what lua offers for concurrency.

kdmcser replied:
how many threads will game use for damage calculator？

IvanSavenko replied:
Why you're asking? 
As many as player has cores in CPU, possibly more.

IvanSavenko replied:
Might actually be interesting idea - allocate N states, and then search for a free one. But probably overkill, at least for now.

So let's start with simple and straightforward solution - protect this class with mutex

kdmcser replied:
in that solution, should I merge setVariable() and evaluate() into one function?

IvanSavenko replied:
Yeah, makes sense

kdmcser replied:
ok

IvanSavenko commented:
@kambala-decapitator 
What should we do with lua library in conan?

Should we wait for conan2 or might be better to rebuild existing conan1 package and add luajit to it?

IvanSavenko commented:
In that case - I'd rather keep this PR open until we can use lua on all platforms, to avoid mods that only work on some systems.

And if luajit can't be used on some platforms - consider how to handle those, like switching to pure lua as fallback

kdmcser commented:
If we cannot import the LuaJIT library, can we manually implement a simple Lua expression parser? I used DeepSeek to generate the parsing code:

`#include <string>
#include <memory>
#include <cmath>
#include <stdexcept>
#include <cctype>
#include <unordered_set>
#include <unordered_map>

using namespace std;

enum class TokenType {
    Number, Variable, Plus, Minus, Multiply, Divide, Modulo, Power, LeftParen, RightParen, End, Invalid
};

struct Token {
    TokenType type;
    string lexeme;
    double numberValue;

    Token(TokenType t, const string& l, double num) 
        : type(t), lexeme(l), numberValue(num) {}
};

class Lexer {
    string input;
    size_t position = 0;

    void skipWhitespace() {
        while (position < input.size() && isspace(input[position])) ++position;
    }

    Token parseNumber() {
        size_t start = position;
        bool hasDot = false;
        while (position < input.size()) {
            char c = input[position];
            if (isdigit(c)) ++position;
            else if (c == '.' && !hasDot) { hasDot = true; ++position; }
            else break;
        }
        string numStr = input.substr(start, position - start);
        try {
            return {TokenType::Number, numStr, stod(numStr)};
        } catch (...) {
            return {TokenType::Invalid, numStr, 0};
        }
    }

    Token parseIdentifier() {
        size_t start = position;
        while (position < input.size() && isalnum(input[position])) ++position;
        string id = input.substr(start, position - start);
        return {TokenType::Variable, id, 0}; // 接受所有合法标识符作为变量
    }

public:
    Lexer(const string& input) : input(input) {}

    Token getNextToken() {
        skipWhitespace();
        if (position >= input.size()) return {TokenType::End, "", 0};

        char c = input[position];
        if (isdigit(c) || c == '.') return parseNumber();
        if (isalpha(c)) return parseIdentifier();

        switch (c) {
            case '+': ++position; return {TokenType::Plus, "+", 0};
            case '-': ++position; return {TokenType::Minus, "-", 0};
            case '*': ++position; return {TokenType::Multiply, "*", 0};
            case '/': ++position; return {TokenType::Divide, "/", 0};
            case '%': ++position; return {TokenType::Modulo, "%", 0};
            case '^': ++position; return {TokenType::Power, "^", 0};
            case '(': ++position; return {TokenType::LeftParen, "(", 0};
            case ')': ++position; return {TokenType::RightParen, ")", 0};
            default: ++position; return {TokenType::Invalid, string(1, c), 0};
        }
    }
};

class ExprNode {
public:
    virtual ~ExprNode() = default;
    virtual double evaluate(const unordered_map<string, double>& vars) const = 0;
};

class BinaryOpNode : public ExprNode {
    char op;
    unique_ptr<ExprNode> left, right;

    static double lua_mod(double a, double b) {
        return a - floor(a / b) * b;
    }

public:
    BinaryOpNode(char op, unique_ptr<ExprNode> l, unique_ptr<ExprNode> r)
        : op(op), left(move(l)), right(move(r)) {}

    double evaluate(const unordered_map<string, double>& vars) const override {
        double lVal = left->evaluate(vars);
        double rVal = right->evaluate(vars);
        switch(op) {
            case '+': return lVal + rVal;
            case '-': return lVal - rVal;
            case '*': return lVal * rVal;
            case '/': 
                if (rVal == 0) throw runtime_error("Division by zero");
                return lVal / rVal;
            case '^': return pow(lVal, rVal);
            case '%': 
                if (rVal == 0) throw runtime_error("Modulo by zero");
                return lua_mod(lVal, rVal);
            default: throw runtime_error("Unknown operator");
        }
    }
};

class UnaryOpNode : public ExprNode {
    char op;
    unique_ptr<ExprNode> operand;

public:
    UnaryOpNode(char op, unique_ptr<ExprNode> node) 
        : op(op), operand(move(node)) {}

    double evaluate(const unordered_map<string, double>& vars) const override {
        double val = operand->evaluate(vars);
        return op == '-' ? -val : val;
    }
};

class VariableNode : public ExprNode {
    string name;

public:
    VariableNode(const string& name) : name(name) {}

    double evaluate(const unordered_map<string, double>& vars) const override {
        auto it = vars.find(name);
        if (it == vars.end()) {
            throw runtime_error("Undefined variable: " + name);
        }
        return it->second;
    }
};

class NumberNode : public ExprNode {
    double value;

public:
    NumberNode(double value) : value(value) {}

    double evaluate(const unordered_map<string, double>&) const override {
        return value;
    }
};

class Parser {
    Lexer& lexer;
    Token currentToken;
    unordered_set<string> allowedVariables;

    void eat(TokenType expected) {
        if (currentToken.type == expected) {
            currentToken = lexer.getNextToken();
        } else {
            throw runtime_error("Syntax error");
        }
    }

    unique_ptr<ExprNode> parsePrimary() {
        if (currentToken.type == TokenType::Number) {
            double val = currentToken.numberValue;
            eat(TokenType::Number);
            return make_unique<NumberNode>(val);
        }

        if (currentToken.type == TokenType::Variable) {
            string name = currentToken.lexeme;
            if (allowedVariables.find(name) == allowedVariables.end()) {
                throw runtime_error("Undefined variable: " + name);
            }
            eat(TokenType::Variable);
            return make_unique<VariableNode>(name);
        }

        if (currentToken.type == TokenType::LeftParen) {
            eat(TokenType::LeftParen);
            auto node = parseExpression();
            eat(TokenType::RightParen);
            return node;
        }

        if (currentToken.type == TokenType::Minus) {
            eat(TokenType::Minus);
            return make_unique<UnaryOpNode>('-', parsePrimary());
        }

        throw runtime_error("Unexpected token");
    }

    unique_ptr<ExprNode> parseFactor() {
        auto node = parsePrimary();
        while (currentToken.type == TokenType::Power) {
            eat(TokenType::Power);
            node = make_unique<BinaryOpNode>('^', move(node), parseFactor());
        }
        return node;
    }

    unique_ptr<ExprNode> parseTerm() {
        auto node = parseFactor();
        while (currentToken.type == TokenType::Multiply || 
               currentToken.type == TokenType::Divide ||
               currentToken.type == TokenType::Modulo) {
            char op = currentToken.lexeme[0];
            eat(currentToken.type);
            node = make_unique<BinaryOpNode>(op, move(node), parseFactor());
        }
        return node;
    }

    unique_ptr<ExprNode> parseExpression() {
        auto node = parseTerm();
        while (currentToken.type == TokenType::Plus || 
               currentToken.type == TokenType::Minus) {
            char op = currentToken.lexeme[0];
            eat(currentToken.type);
            node = make_unique<BinaryOpNode>(op, move(node), parseTerm());
        }
        return node;
    }

public:
    Parser(Lexer& lexer, const unordered_set<string>& vars) 
        : lexer(lexer), allowedVariables(vars) {
        currentToken = lexer.getNextToken();
    }

    unique_ptr<ExprNode> parse() {
        auto expr = parseExpression();
        if (currentToken.type != TokenType::End) {
            throw runtime_error("Unexpected end of expression");
        }
        return expr;
    }
};

class LuaExpressionParser {
    unique_ptr<ExprNode> ast;
    unordered_set<string> allowedVariables;

    string preprocess(const string& expr) {
        const string prefix = "return ";
        if (expr.substr(0, prefix.size()) == prefix) {
            return expr.substr(prefix.size());
        }
        return expr;
    }

public:
    LuaExpressionParser(const string& expr, const vector<string>& variables) {
        for (const auto& var : variables) {
            allowedVariables.insert(var);
        }
        string processed = preprocess(expr);
        Lexer lexer(processed);
        Parser parser(lexer, allowedVariables);
        ast = parser.parse();
    }

    double evaluate(const unordered_map<string, double>& variables) {
        return ast->evaluate(variables);
    }
};
`

kambala-decapitator commented:
we can import it, just wait a bit please

Opuszek committed:
fix third dragon slayer mission

IvanSavenko committed:
Switch lobby from `beholder.vcmi.eu` to `lobby.vcmi.eu`

IvanSavenko committed:
Add time tracker to all lobby database calls

IvanSavenko committed:
Add existing lobby database schema to config

IvanSavenko committed:
Remove no longer relevant code, add missing indexes to SQL schema

IvanSavenko committed:
Add untested DB upgrade script

IvanSavenko committed:
Generate .deb archive as part of lobby test CI run

IvanSavenko committed:
Mods active in room are now stored in separate table

IvanSavenko committed:
Better db upgrade script

IvanSavenko committed:
Fix bugs discovered during tests

IvanSavenko committed:
Do not abort lobby server on database access from another process

IvanSavenko committed:
Update docs

IvanSavenko committed:
API docs should now match actual API reply

IvanSavenko committed:
Only send update on a single room when romm is changed

IvanSavenko committed:
Fix issues found in review

IvanSavenko committed:
Dump performance metrics daily

IvanSavenko committed:
Update docs

IvanSavenko committed:
Attach instance bonuses to newly assembled combined arts

IvanSavenko committed:
Exclude war machines from mods from vcmiartifacts cheat

IvanSavenko committed:
Added warning on attempt to replace war machine in Blacksmith

IvanSavenko committed:
Use correct singular / plural form in creature window

IvanSavenko committed:
Do not show option to shoot for blocked units

IvanSavenko committed:
Fix loading of vcmiserver port in Launcher settings

IvanSavenko committed:
Correctly reset translation to English when requested

IvanSavenko committed:
Fix spell icon size & Shrine hover text to match H3

IvanSavenko committed:
Fix swapped messages in Altar

Laserlicht committed:
add new buttons

Laserlicht committed:
Merge branch 'beta' into advopt

Laserlicht committed:
codestyle

Laserlicht committed:
add modscope

Laserlicht committed:
fixes

GeorgeK1ng committed:
Support for generating dialogs with status bar

GeorgeK1ng committed:
Backpack status bar text fix

GeorgeK1ng committed:
Start game improvements

GeorgeK1ng committed:
Improve multiplayer selection dialog

GeorgeK1ng committed:
Fix random map generating

GeorgeK1ng committed:
Fix maps filtering performance issue

Opuszek committed:
Adds battle event to walk and spellcast action

kambala-decapitator committed:
end turn check: slight refactoring

kambala-decapitator committed:
clean vertical space

kambala-decapitator committed:
fix check to show the move reminder dialog

when hero can still move and can move along the planned path

GeorgeK1ng committed:
Refactor save and load methods in CGameHandler

Karyoplasma committed:
fix missing quote and add default sizes

IvanSavenko committed:
Allow 3 players on Coldshadow Fantasy template to cover missing case

IvanSavenko committed:
Fix formatting

IvanSavenko committed:
War Machine Factory now also warns on replacing war machine

IvanSavenko committed:
Merge pull request #7269 from Karyoplasma/artifact_format

[Documentation] Fix missing quote and add default sizes in Artifact format

IvanSavenko committed:
Merge pull request #7266 from Opuszek/make_walk_and_spellcast_trigger_battle_event

[1.7.4]  Adds battle event to walk and spellcast action

IvanSavenko committed:
Merge pull request #7268 from GeorgeK1ng/load_save_fix

[1.7.4] Fix save load

IvanSavenko committed:
Merge pull request #7267 from kambala-decapitator/fix-end-turn-dialog

fix check to show the move reminder dialog

IvanSavenko committed:
Merge pull request #7242 from IvanSavenko/bugfixing

[1.7.4] Fixes for reported issues

IvanSavenko committed:
Changelog for 1.7.4

GeorgeK1ng committed:
Remove QStandardPaths::DocumentsLocation usage on VCMI_MOBILE

GeorgeK1ng committed:
ReRun CI

GeorgeK1ng committed:
Improve status bar color

GeorgeK1ng committed:
Make custom size dialog title yellow as other titles

GeorgeK1ng committed:
Reorder host and join buttons in main menu

GeorgeK1ng committed:
Generate University background

Laserlicht committed:
re use strings (code review)

kambala-decapitator committed:
[cmake][macOS] write `LSMinimumSystemVersion` to Info.plist

only when `CMAKE_OSX_DEPLOYMENT_TARGET` is set

GeorgeK1ng committed:
Merge pull request #7204 from GeorgeK1ng/long_weapon

[1.7.4?] Add LONG_WEAPON bonus

IvanSavenko committed:
Merge pull request #7272 from GeorgeK1ng/documents_folder

[1.7.4] Prevents Documents folder creation on VCMI_MOBILE

GeorgeK1ng committed:
Add Spell Research background

GeorgeK1ng committed:
Improve Spell Research button spacing

GeorgeK1ng committed:
RaRun CI

Opuszek committed:
Adds disposed heroes selector to map editor

Laserlicht committed:
code review points

Laserlicht committed:
remaining code review points

Laserlicht committed:
allow to open wiki from main menu (needs custom main menu)

Laserlicht committed:
list layout double height

Laserlicht committed:
layout

Laserlicht committed:
add mod page

IvanSavenko committed:
Merge pull request #7274 from kambala-decapitator/macos-min-version-info-plist

[cmake][macOS] write deployment target to Info.plist

IvanSavenko committed:
Merge pull request #7259 from Laserlicht/advopt

[1.7.4] Adventure Options scrollable and extendable

Laserlicht committed:
Merge branch 'beta' into wiki

Laserlicht committed:
wiki adventure option

GeorgeK1ng committed:
Merge branch 'dialog_with_status' of https://github.com/GeorgeK1ng/vcmi into dialog_with_status

GeorgeK1ng committed:
Simplify player colored frame with status bar generating

GeorgeK1ng committed:
Apply forgotten change for Spell research

GeorgeK1ng committed:
Fix Hero Window status bar

Laserlicht committed:
use icons as table header in wiki

Laserlicht committed:
panning fix

Laserlicht committed:
small text fix

Laserlicht committed:
smooth scroll in cviewport for panning

Laserlicht committed:
fix text bounding

Laserlicht committed:
fix redraw

Opuszek committed:
Fixes a bug where a disposed hero could be chosen as the random starting hero

Laserlicht committed:
disable hover effect for panning

Laserlicht committed:
allow Inertia also for gamelist and objectlist

Laserlicht committed:
add Inertia for missing wiki lists

IvanSavenko committed:
Fix crash on purchasing war machine when slot is empty

IvanSavenko committed:
Merge pull request #7260 from GeorgeK1ng/dialog_with_status

[1.7.4] Generate background for Recruitment, University, Backpack and Spell Research dialogs

IvanSavenko committed:
Merge pull request #7277 from IvanSavenko/war_machine_fix

Fix crash on purchasing war machine when slot is empty

GeorgeK1ng committed:
Generate Stack Experience Detail window

IvanSavenko committed:
Merge pull request #6920 from Opuszek/fix_third_dragon_slayer_mission

fix third dragon slayer mission

IvanSavenko committed:
Auto-update VCMI Qt translation files

IvanSavenko committed:
Merge pull request #7275 from Opuszek/map_editor_add_disposed_heroes_selector

Map editor add disposed heroes selector

IvanSavenko committed:
Merge pull request #7223 from Laserlicht/wiki

[1.7.4] ingame wiki

Laserlicht committed:
Fix asset gen

IvanSavenko committed:
Merge pull request #7287 from vcmi/fix_asset_gen

[1.7.4] Fix asset gen

Opuszek committed:
MapEditor adds portraits to hero settings

Laserlicht committed:
improve highlighted state for selected wiki element

Laserlicht committed:
support for custom description/icon in wiki bonus

Laserlicht committed:
markdown parser; fix center (entries); fix highlight

Laserlicht committed:
allow to load english strings, when there is no translated ones

Laserlicht committed:
implement anchors for wiki

Laserlicht committed:
fix docs

Laserlicht committed:
changed to dict to allow assembleFromFIle

Laserlicht committed:
possiblity to define own categories

Laserlicht committed:
no sorting for custom categories

Laserlicht committed:
add possibility to do cutom order

Laserlicht committed:
fix docs linter error

Laserlicht committed:
language fallback fix (when partially translated)

Laserlicht committed:
blockquotes, text styles and code(blocks)

Laserlicht committed:
fix translation edge case

Laserlicht committed:
add movement costs and spell school

Laserlicht committed:
refactor to avoid copied code

Laserlicht committed:
fix inline line break

Laserlicht committed:
refactor to seperate functions

Laserlicht committed:
make sonarcloud happy

Laserlicht committed:
fixed sonarcloud and horicontal scrollbar

IvanSavenko committed:
Translated using Weblate (Ukrainian)

Currently translated at 100.0% (1105 of 1105 strings)

Translated using Weblate (Ukrainian)

Currently translated at 100.0% (1099 of 1099 strings)

Translated using Weblate (Ukrainian)

Currently translated at 100.0% (1099 of 1099 strings)

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/uk/
Translation: VCMI/Game client

XCOM-HUB committed:
Translated using Weblate (Swedish)

Currently translated at 100.0% (416 of 416 strings)

Translated using Weblate (Swedish)

Currently translated at 100.0% (1179 of 1179 strings)

Translated using Weblate (Swedish)

Currently translated at 100.0% (925 of 925 strings)

Translated using Weblate (Swedish)

Currently translated at 100.0% (1105 of 1105 strings)

Translated using Weblate (Swedish)

Currently translated at 100.0% (1099 of 1099 strings)

Translated using Weblate (Swedish)

Currently translated at 100.0% (925 of 925 strings)

Translated using Weblate (Swedish)

Currently translated at 100.0% (416 of 416 strings)

Co-authored-by: Maurycy <maurycy.domaradzki@outlook.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/sv/
Translate-URL: https://hosted.weblate.org/projects/vcmi/editor/sv/
Translate-URL: https://hosted.weblate.org/projects/vcmi/launcher/sv/
Translation: VCMI/Game client
Translation: VCMI/Launcher
Translation: VCMI/Map Editor

cyphra committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (1099 of 1099 strings)

Co-authored-by: Francisco Serrador <fserrador@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/es/
Translation: VCMI/Game client

GeorgeK1ng committed:
Translated using Weblate (Czech)

Currently translated at 98.7% (411 of 416 strings)

Translated using Weblate (Czech)

Currently translated at 100.0% (1099 of 1099 strings)

Translated using Weblate (Czech)

Currently translated at 100.0% (925 of 925 strings)

Co-authored-by: George King <George.King@seznam.cz>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/cs/
Translate-URL: https://hosted.weblate.org/projects/vcmi/editor/cs/
Translate-URL: https://hosted.weblate.org/projects/vcmi/launcher/cs/
Translation: VCMI/Game client
Translation: VCMI/Launcher
Translation: VCMI/Map Editor

Commit:
Translated using Weblate (Japanese)

Currently translated at 97.0% (1069 of 1102 strings)

Translated using Weblate (Japanese)

Currently translated at 96.9% (1066 of 1099 strings)

Co-authored-by: Carson_N <sakai46jp@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ja/
Translation: VCMI/Game client

Laserlicht committed:
Translated using Weblate (German)

Currently translated at 100.0% (1179 of 1179 strings)

Translated using Weblate (German)

Currently translated at 100.0% (1099 of 1099 strings)

Co-authored-by: Laserlicht <13953785+Laserlicht@users.noreply.github.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/de/
Translation: VCMI/Game client

Karyoplasma committed:
Translated using Weblate (German)

Currently translated at 100.0% (1105 of 1105 strings)

Translated using Weblate (German)

Currently translated at 100.0% (416 of 416 strings)

Translated using Weblate (German)

Currently translated at 100.0% (1099 of 1099 strings)

Co-authored-by: Karyoplasma <karyoplasma@proton.me>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/de/
Translate-URL: https://hosted.weblate.org/projects/vcmi/launcher/de/
Translation: VCMI/Game client
Translation: VCMI/Launcher

Keloraen committed:
Translated using Weblate (Russian)

Currently translated at 88.2% (970 of 1099 strings)

Co-authored-by: Konrad Donner <konrad800@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ru/
Translation: VCMI/Game client

IvanSavenko committed:
Merge pull request #7285 from vcmi/bot/qt-translation-refresh/beta

Auto-update Qt translation files

IvanSavenko committed:
Merge pull request #7233 from vcmi/weblate

Translations update from Hosted Weblate

GeorgeK1ng committed:
Merge branch 'beta' into dialog_with_status

IvanSavenko committed:
Better translation exporting for mods

Better version of translation exporting logic. Compared to existiing
version it:
- places generated json's in same directory structure as recommended for
mods (`modname/Content/configtranslation/language.json`). Files are
placed in same directory before (`exported`) to reduce chance of
information loss on overwrite
- (mostly) correctly handled mods that overwrite strings from another
submod of the same mod. For now only simple cases are handled (within
same mod, and without long overwrite chains), which seems to be
sufficient for existing mods

New translation is done by server (vcmiserver / VCMI_Server.exe) and not
by client command - this is due to reloading of library in runtime which
at the moment can't be done on client, especially during ongoing game

Laserlicht committed:
refactored; removed some markdown functionality

Laserlicht committed:
remove leftover

Laserlicht committed:
last leftover

GeorgeK1ng committed:
Make Stack Experience Details Window dynamic and standalone

IvanSavenko committed:
Use strict json validation for translations

IvanSavenko committed:
Export campaigns and maps in mod that added map/campaign

GeorgeK1ng committed:
Fix missing class includes

GeorgeK1ng committed:
Fix background table data rows

IvanSavenko committed:
Add weblate docs

GeorgeK1ng committed:
Remove unneded strings, fix stack bonuses

GeorgeK1ng committed:
Fix generated icons, fix bonuses, add missing strings

GeorgeK1ng committed:
Language strings cleanup, correct stack experience bonus source indicator

IvanSavenko committed:
Merge pull request #7288 from Opuszek/me_add_images_to_disposed_heroes_selection

MapEditor adds portraits to hero settings

IvanSavenko committed:
Update docs

IvanSavenko committed:
Fix regressions

IvanSavenko committed:
Merge pull request #7149 from IvanSavenko/lobby_upgrade

Lobby server upgrade and migration preparation

IvanSavenko committed:
Merge pull request #7293 from IvanSavenko/better_translate

[1.7.4?] Better translation exporting for mods

GeorgeK1ng committed:
Address comments

GeorgeK1ng committed:
Leftover removal

IvanSavenko committed:
Merge pull request #7279 from GeorgeK1ng/dialog_with_status

[1.7.4] Generate Creature Stack Experience Detail window

GeorgeK1ng committed:
Merge branch 'beta' into game_start

GeorgeK1ng committed:
Fix Begin button blink, statusbar and Sonar

IvanSavenko committed:
Merge pull request #7289 from vcmi/wiki_improve

[1.7.4] Wiki improve

weblate committed:
Merge pull request #7292 from vcmi/weblate

Translations update from Hosted Weblate

GeorgeK1ng committed:
Fix datarowws logic + Experience description

Laserlicht committed:
fix panning for wiki

Laserlicht committed:
improve scrollbar thumb

IvanSavenko committed:
Fixes for lobby regressions

- Fixed client not handling updateGameRoom message for new rooms
- Fixed account ID not updating correctly when using account details
from beholder for lobby domain name
- Removed unnecessary sending of all game rooms when server starts
connection. Now list is only send once connection is done
- Reverted some changes for 1.7.3 compatibility

IvanSavenko committed:
Merge pull request #7301 from Laserlicht/fixes

[1.7.4] small fixes related to scrolling

IvanSavenko committed:
Merge pull request #7304 from IvanSavenko/lobby_fix

[1.7.4] Fixes for lobby regressions

IvanSavenko committed:
Merge pull request #7271 from IvanSavenko/changelog_174

Changelog for 1.7.4

Laserlicht committed:
wiki improvements

Laserlicht committed:
remove not needed includes

Laserlicht committed:
more places to open wiki

Laserlicht committed:
fix history

IvanSavenko committed:
Workaround for loading vmap's with invalid spells

IvanSavenko committed:
Merge pull request #7300 from GeorgeK1ng/exp_fix

[1.7.4] Improve Stack Experience Details

IvanSavenko committed:
Optimize getActiveGameRooms call

This call currently takes large part of DB access time.

According to EXPLAIN QUERY PLAN it looks like WHERE xxx IN(1,2,3) causes
SQLite to use full table SCAN instead of using available index.

For getActiveGameRooms this change decreases query time from ~60-70 ms
to <1 ms

IvanSavenko committed:
Merge pull request #7306 from IvanSavenko/vmap_fix

Workaround for loading vmap's with invalid spells

IvanSavenko committed:
Merge pull request #7307 from IvanSavenko/db_fix

Optimize getActiveGameRooms call

Laserlicht committed:
add statistic button

GeorgeK1ng committed:
Fix Start button handling for all multiplayer games

GeorgeK1ng committed:
ReRun CI

IvanSavenko committed:
Merge pull request #7310 from Laserlicht/button

[1.7.4] add statistic button

GeorgeK1ng committed:
Merge pull request #7311 from GeorgeK1ng/map_format

[1.7.4] Improve WoG map support

IvanSavenko committed:
[1.7.4] Wiki improvements (#7305)

* wiki improvements

* remove not needed includes

* more places to open wiki

* fix history

GeorgeK1ng committed:
Apply code review and simplify

GeorgeK1ng committed:
Fix Sonar warning

Laserlicht committed:
added missing translation

GeorgeK1ng committed:
Fix another Sonar warning

IvanSavenko committed:
Merge pull request #7314 from Laserlicht/trans

[1.7.4] added missing translation string

GeorgeK1ng committed:
Code review

GeorgeK1ng committed:
Fix compile

IvanSavenko committed:
Merge pull request #7262 from GeorgeK1ng/game_start

[1.7.4] Various game start improvements

IvanSavenko committed:
Release preparation for 1.7.4

- Updated changelog with final changes
- Changed Android build ID to next available
- Updated Readme to include new release counter
- Set release date to 17th May

Changes in code
- Fixed possible crash in high scores on having two creatures with same
ID from different mods

Laserlicht committed:
fix scroll

Laserlicht committed:
fixes difficulty icons click when in chat

Laserlicht committed:
more places for inertia scroll

Laserlicht committed:
fix flickering number when changing camp bonus selection

Laserlicht committed:
reset filter when navigate in wiki

IvanSavenko committed:
Merge pull request #7324 from Laserlicht/fixes_174

[1.7.4] small fixes and improvements for 1.7.4

IvanSavenko committed:
Merge pull request #7322 from Laserlicht/scroll_fix

[1.7.4] fix scrolling on wayland

weblate committed:
Merge pull request #7303 from vcmi/weblate

Translations update from Hosted Weblate

IvanSavenko committed:
Merge pull request #7316 from IvanSavenko/release_174

Release preparation for 1.7.4

IvanSavenko committed:
Translated using Weblate (Belarusian)

Currently translated at 0.0% (0 of 18 strings)

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/be/
Translation: VCMI/Linux metadata

weblate committed:
Translated using Weblate (Norwegian Bokmål)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (French)

Currently translated at 11.2% (136 of 1206 strings)

Translated using Weblate (Swedish)

Currently translated at 98.5% (1188 of 1206 strings)

Translated using Weblate (Belarusian)

Currently translated at 0.0% (0 of 18 strings)

Co-authored-by: Anonymous <noreply@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/nb_NO/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/sv/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/be/
Translation: VCMI/Android
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

SpiritOvRebellion committed:
Translated using Weblate (Dutch)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (Romanian)

Currently translated at 74.7% (902 of 1206 strings)

Translated using Weblate (French)

Currently translated at 11.2% (136 of 1206 strings)

Co-authored-by: Nowy dzień <literad@protonmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/nl/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ro/
Translation: VCMI/Game client

Karyoplasma committed:
Translated using Weblate (Dutch)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (Romanian)

Currently translated at 74.7% (902 of 1206 strings)

Translated using Weblate (French)

Currently translated at 11.2% (136 of 1206 strings)

Co-authored-by: Karyoplasma <karyoplasma@proton.me>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/nl/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ro/
Translation: VCMI/Game client

IvanSavenko committed:
Translated using Weblate (Dutch)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (Romanian)

Currently translated at 74.7% (902 of 1206 strings)

Translated using Weblate (Ukrainian)

Currently translated at 100.0% (18 of 18 strings)

Translated using Weblate (French)

Currently translated at 11.2% (136 of 1206 strings)

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/nl/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ro/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/uk/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

weblate committed:
Translated using Weblate (Bulgarian)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Finnish)

Currently translated at 0.0% (0 of 1206 strings)

Translated using Weblate (Dutch)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (Romanian)

Currently translated at 74.7% (902 of 1206 strings)

Translated using Weblate (Vietnamese)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Ukrainian)

Currently translated at 100.0% (18 of 18 strings)

Co-authored-by: Anonymous <noreply@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/bg/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/vi/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fi/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/nl/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ro/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/uk/
Translation: VCMI/Android
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

SpiritOvRebellion committed:
Translated using Weblate (Latvian)

Currently translated at 79.6% (960 of 1206 strings)

Translated using Weblate (Vietnamese)

Currently translated at 78.6% (949 of 1206 strings)

Translated using Weblate (Finnish)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Nowy dzień <literad@protonmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fi/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/lv/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/vi/
Translation: VCMI/Game client

Karyoplasma committed:
Translated using Weblate (Vietnamese)

Currently translated at 78.6% (949 of 1206 strings)

Translated using Weblate (Finnish)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Karyoplasma <karyoplasma@proton.me>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fi/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/vi/
Translation: VCMI/Game client

IvanSavenko committed:
Translated using Weblate (Polish)

Currently translated at 100.0% (18 of 18 strings)

Translated using Weblate (Turkish)

Currently translated at 100.0% (18 of 18 strings)

Translated using Weblate (Japanese)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Serbian)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Chinese (Simplified Han script))

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Latvian)

Currently translated at 79.6% (960 of 1206 strings)

Translated using Weblate (Russian)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Norwegian Bokmål)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Vietnamese)

Currently translated at 78.6% (949 of 1206 strings)

Translated using Weblate (Italian)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Finnish)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fi/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/lv/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/vi/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/it/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/ja/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/nb_NO/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/pl/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/ru/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/sr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/tr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/zh_Hans/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

weblate committed:
Translated using Weblate (Hungarian)

Currently translated at 45.0% (543 of 1206 strings)

Translated using Weblate (Korean)

Currently translated at 0.0% (0 of 1206 strings)

Translated using Weblate (Latvian)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Portuguese)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Chinese (Simplified Han script))

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Japanese)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Serbian)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Italian)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Filipino)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Finnish)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Japanese)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Chinese (Simplified Han script))

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Latvian)

Currently translated at 79.6% (960 of 1206 strings)

Translated using Weblate (Romanian)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Russian)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Norwegian Bokmål)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Vietnamese)

Currently translated at 78.6% (949 of 1206 strings)

Translated using Weblate (Italian)

Currently translated at 0.0% (0 of 18 strings)

Co-authored-by: Anonymous <noreply@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/fi/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/fil/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/it/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/ja/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/lv/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/pt/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/ro/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/zh_Hans/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/hu/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ko/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/lv/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/vi/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/it/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/ja/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/nb_NO/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/ru/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/sr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/zh_Hans/
Translation: VCMI/Android
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

Erthon committed:
Translated using Weblate (Turkish)

Currently translated at 100.0% (18 of 18 strings)

Co-authored-by: Ertuğrul Hazar <ertugrulhazar@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/tr/
Translation: VCMI/Linux metadata

SpiritOvRebellion committed:
Translated using Weblate (Serbian)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (Portuguese)

Currently translated at 48.0% (579 of 1206 strings)

Translated using Weblate (Hungarian)

Currently translated at 45.0% (543 of 1206 strings)

Translated using Weblate (Korean)

Currently translated at 0.0% (0 of 1206 strings)

Translated using Weblate (Polish)

Currently translated at 100.0% (18 of 18 strings)

Co-authored-by: Nowy dzień <literad@protonmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/hu/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ko/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/pt/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/sr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/pl/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

Karyoplasma committed:
Translated using Weblate (Hungarian)

Currently translated at 45.0% (543 of 1206 strings)

Translated using Weblate (Korean)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Karyoplasma <karyoplasma@proton.me>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/hu/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ko/
Translation: VCMI/Game client

IvanSavenko committed:
Translated using Weblate (Serbian)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (French)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Greek)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Portuguese)

Currently translated at 48.0% (579 of 1206 strings)

Translated using Weblate (Romanian)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Hungarian)

Currently translated at 45.0% (543 of 1206 strings)

Translated using Weblate (Korean)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/hu/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ko/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/pt/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/sr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/el/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/fr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/ro/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

weblate committed:
Translated using Weblate (Serbian)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (French)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Greek)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Portuguese)

Currently translated at 48.0% (579 of 1206 strings)

Translated using Weblate (Romanian)

Currently translated at 0.0% (0 of 18 strings)

Co-authored-by: Anonymous <noreply@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/pt/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/sr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/el/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/fr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/ro/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

Karyoplasma committed:
Translated using Weblate (German)

Currently translated at 100.0% (18 of 18 strings)

Translated using Weblate (Serbian)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (Portuguese)

Currently translated at 48.0% (579 of 1206 strings)

Co-authored-by: Karyoplasma <karyoplasma@proton.me>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/pt/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/sr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/de/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

weblate committed:
Translated using Weblate (Filipino)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (Bulgarian)

Currently translated at 0.0% (0 of 1206 strings)

Translated using Weblate (Spanish)

Currently translated at 90.0% (1086 of 1206 strings)

Translated using Weblate (German)

Currently translated at 100.0% (18 of 18 strings)

Co-authored-by: Anonymous <noreply@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/bg/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/es/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fil/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/de/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

SpiritOvRebellion committed:
Translated using Weblate (Filipino)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (Bulgarian)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Nowy dzień <literad@protonmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/bg/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fil/
Translation: VCMI/Game client

Karyoplasma committed:
Translated using Weblate (Filipino)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (Bulgarian)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Karyoplasma <karyoplasma@proton.me>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/bg/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fil/
Translation: VCMI/Game client

IvanSavenko committed:
Translated using Weblate (Hungarian)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Filipino)

Currently translated at 79.1% (955 of 1206 strings)

Translated using Weblate (Bulgarian)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/bg/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/fil/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/hu/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

weblate committed:
Translated using Weblate (Belarusian)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Serbian)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Italian)

Currently translated at 45.1% (545 of 1206 strings)

Translated using Weblate (Korean)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Hungarian)

Currently translated at 0.0% (0 of 18 strings)

Co-authored-by: Anonymous <noreply@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/be/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/ko/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/sr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/it/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/hu/
Translation: VCMI/Android
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

SpiritOvRebellion committed:
Translated using Weblate (Italian)

Currently translated at 45.1% (545 of 1206 strings)

Co-authored-by: Nowy dzień <literad@protonmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/it/
Translation: VCMI/Game client

Karyoplasma committed:
Translated using Weblate (Italian)

Currently translated at 45.1% (545 of 1206 strings)

Co-authored-by: Karyoplasma <karyoplasma@proton.me>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/it/
Translation: VCMI/Game client

IvanSavenko committed:
Translated using Weblate (Vietnamese)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Finnish)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Filipino)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Italian)

Currently translated at 45.1% (545 of 1206 strings)

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/it/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/fi/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/fil/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/vi/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

weblate committed:
Translated using Weblate (Chinese (Simplified Han script))

Currently translated at 79.9% (964 of 1206 strings)

Translated using Weblate (Russian)

Currently translated at 50.0% (1 of 2 strings)

Translated using Weblate (Norwegian Bokmål)

Currently translated at 0.0% (0 of 1206 strings)

Translated using Weblate (Vietnamese)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Finnish)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Filipino)

Currently translated at 0.0% (0 of 18 strings)

Co-authored-by: Anonymous <noreply@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/ru/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/nb_NO/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/zh_Hans/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/fi/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/fil/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/vi/
Translation: VCMI/Android
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

SpiritOvRebellion committed:
Translated using Weblate (Norwegian Bokmål)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Nowy dzień <literad@protonmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/nb_NO/
Translation: VCMI/Game client

Karyoplasma committed:
Translated using Weblate (Norwegian Bokmål)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Karyoplasma <karyoplasma@proton.me>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/nb_NO/
Translation: VCMI/Game client

IvanSavenko committed:
Translated using Weblate (Latvian)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Chinese (Simplified Han script))

Currently translated at 79.9% (964 of 1206 strings)

Translated using Weblate (Norwegian Bokmål)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/nb_NO/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/zh_Hans/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/lv/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

weblate committed:
Translated using Weblate (Belarusian)

Currently translated at 0.0% (0 of 1206 strings)

Translated using Weblate (Latvian)

Currently translated at 0.0% (0 of 18 strings)

Co-authored-by: Anonymous <noreply@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/be/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/lv/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

SpiritOvRebellion committed:
Translated using Weblate (Russian)

Currently translated at 79.4% (958 of 1206 strings)

Translated using Weblate (Belarusian)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Nowy dzień <literad@protonmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/be/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ru/
Translation: VCMI/Game client

Karyoplasma committed:
Translated using Weblate (Belarusian)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Karyoplasma <karyoplasma@proton.me>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/be/
Translation: VCMI/Game client

IvanSavenko committed:
Translated using Weblate (Swedish)

Currently translated at 100.0% (18 of 18 strings)

Translated using Weblate (Russian)

Currently translated at 79.4% (958 of 1206 strings)

Translated using Weblate (Bulgarian)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Belarusian)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/be/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ru/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/bg/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/sv/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

weblate committed:
Translated using Weblate (Greek)

Currently translated at 0.0% (0 of 1206 strings)

Translated using Weblate (Russian)

Currently translated at 79.4% (958 of 1206 strings)

Translated using Weblate (Bulgarian)

Currently translated at 0.0% (0 of 18 strings)

Co-authored-by: Anonymous <noreply@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/el/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/ru/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/bg/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

XCOM-HUB committed:
Translated using Weblate (Swedish)

Currently translated at 100.0% (18 of 18 strings)

Co-authored-by: Maurycy <maurycy.domaradzki@outlook.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/sv/
Translation: VCMI/Linux metadata

SpiritOvRebellion committed:
Translated using Weblate (Greek)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Nowy dzień <literad@protonmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/el/
Translation: VCMI/Game client

Karyoplasma committed:
Translated using Weblate (Greek)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Karyoplasma <karyoplasma@proton.me>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/el/
Translation: VCMI/Game client

IvanSavenko committed:
Translated using Weblate (Korean)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Portuguese)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Dutch)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Greek)

Currently translated at 0.0% (0 of 1206 strings)

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/el/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/ko/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/nl/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/pt/
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

weblate committed:
Translated using Weblate (Korean)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (Turkish)

Currently translated at 88.7% (1070 of 1206 strings)

Translated using Weblate (Hungarian)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Portuguese)

Currently translated at 0.0% (0 of 18 strings)

Translated using Weblate (French)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Dutch)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Dutch)

Currently translated at 0.0% (0 of 18 strings)

Co-authored-by: Anonymous <noreply@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/fr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/hu/
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/nl/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/tr/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/ko/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/nl/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/pt/
Translation: VCMI/Android
Translation: VCMI/Game client
Translation: VCMI/Linux metadata

Commit:
Translated using Weblate (Spanish)

Currently translated at 100.0% (18 of 18 strings)

Co-authored-by: Deleted User <noreply+36847@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/es/
Translation: VCMI/Linux metadata

GeorgeK1ng committed:
Translated using Weblate (Czech)

Currently translated at 100.0% (18 of 18 strings)

Co-authored-by: George King <George.King@seznam.cz>
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/cs/
Translation: VCMI/Linux metadata

weblate committed:
Translated using Weblate (Greek)

Currently translated at 0.0% (0 of 2 strings)

Translated using Weblate (Czech)

Currently translated at 100.0% (18 of 18 strings)

Co-authored-by: Anonymous <noreply@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/el/
Translate-URL: https://hosted.weblate.org/projects/vcmi/linux/cs/
Translation: VCMI/Android
Translation: VCMI/Linux metadata

weblate committed:
Update translation files

Updated by "Remove blank strings" add-on in Weblate.

Update translation files

Updated by "Remove blank strings" add-on in Weblate.

Co-authored-by: Hosted Weblate <hosted@weblate.org>
Translate-URL: https://hosted.weblate.org/projects/vcmi/android/
Translate-URL: https://hosted.weblate.org/projects/vcmi/client/
Translation: VCMI/Android
Translation: VCMI/Game client

IvanSavenko committed:
Merge pull request #7328 from vcmi/weblate

Translations update from Hosted Weblate

IvanSavenko committed:
Merge pull request #7329 from vcmi/beta

Merge beta -> master

IvanSavenko committed:
Fixes for crashes in 1.7.4 beta

- Fixed crash in MP between 1.7.3 and 1.7.4 on opening exchange window
- Fixed crash if creature with area attack and death blow attacks empty
space (Magog in WoG)
- Fixed crash on formatting of spell effect preview text
- Fixed crash if hero has invalid path (e.g. blocked by another hero)
and player attempts to end turn
- Try to fix crash on opening battle-only mode in some cases(?)

Laserlicht committed:
fix color in template editor

IvanSavenko committed:
Merge pull request #7334 from IvanSavenko/crashfix

Fixes for crashes in 1.7.4 beta

IvanSavenko committed:
Merge pull request #7336 from Laserlicht/fix_tpl

[1.7.4] fix template editor color

IvanSavenko committed:
Merge pull request #7337 from vcmi/beta

Merge beta -> master

kambala-decapitator committed:
update dependencies submodule and docs

IvanSavenko committed:
Merge pull request #7340 from kambala-decapitator/deps-update

update dependencies submodule and docs

IvanSavenko committed:
Merge pull request #7344 from vcmi/beta

Merge beta -> master

kdmcser committed:
merge master
